src/Microsoft.Windows.ComputeVirtualization/HcsNotification.cs

```diff
@@ -0,0 +1,99 @@
+using System;
+using System.Runtime.InteropServices;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Microsoft.Windows.ComputeVirtualization
+{
+    internal class NotificationResult
+    {
+        public AutoResetEvent Waiter;
+        public int Status;
+        public string Data;
+    }
+
+    internal enum HCS_NOTIFICATIONS : uint
+    {
+        HcsNotificationInvalid = 0x00000000,
+
+        /// Notifications for HCS_SYSTEM handles
+        HcsNotificationSystemExited = 0x00000001,
+        HcsNotificationSystemCreateCompleted = 0x00000002,
+        HcsNotificationSystemStartCompleted = 0x00000003,
+
+        /// Notifications for HCS_PROCESS handles
+        HcsNotificationProcessExited = 0x00010000,
+
+        /// Common notifications
+        HcsNotificationServiceDisconnect = 0x01000000,
+
+        /// The upper 4 bits are reserved for flags. See HCS_NOTIFICATION_FLAGS
+        HcsNotificationFlagsReserved = 0xF0000000
+    }
+
+    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+    internal delegate void NotificationCallback(uint notificationType, IntPtr context, int notificationStatus, [MarshalAs(UnmanagedType.LPWStr)] string notificationData);
+
+    internal delegate int RegisterHcsNotificationCallback(IntPtr handle, NotificationCallback callback, IntPtr context, out IntPtr callbackHandle);
+
+    internal delegate int UnregisterHcsNotificationCallback(IntPtr handle);
+
+    internal class HcsNotificationWatcher : IDisposable
+    {
+        private IntPtr _h;
+        private UnregisterHcsNotificationCallback _unreg;
+        private Dictionary<HCS_NOTIFICATIONS, NotificationResult> _n;
+
+        public HcsNotificationWatcher(IntPtr handle, RegisterHcsNotificationCallback register, UnregisterHcsNotificationCallback unregister, HCS_NOTIFICATIONS[] notificationList)
+        {
+            _unreg = unregister;
+
+            _n = new Dictionary<HCS_NOTIFICATIONS, NotificationResult>();
+            foreach (var notificationType in notificationList)
+            {
+                var entry = new NotificationResult();
+                entry.Waiter = new AutoResetEvent(false);
+                _n.Add(notificationType, entry);
+            }
+
+            NotificationCallback callbackFunc = (uint nType, IntPtr ctx, int nStatus, string nData) =>
+            {
+                var key = (HCS_NOTIFICATIONS)nType;
+                if (_n.ContainsKey(key))
+                {
+                    _n[key].Status = nStatus;
+                    if (nData != null)
+                    {
+                        _n[key].Data = new string(nData.ToCharArray());
+                    }
+                    _n[key].Waiter.Set();
+                }
+            };
+
+            HcsFunctions.ProcessHcsCall(register(handle, callbackFunc, IntPtr.Zero, out _h));
+        }
+
+        public bool Wait(HCS_NOTIFICATIONS notificationType, int timeout = Timeout.Infinite)
```

Comment:
Not sure I care but this should likely check if you are waiting on an event your not listening for. At least as a debug trace.